### PR TITLE
Adol-C advanced branching: Fix FEValues and tensor classes

### LINKED ISSUE #6339: Compilation fails with ADOL-C (current master) configured with --enable-advanced-branching
@jppelteret Following your recommendation I have updated the Gentoo ebuild for ADOL-C to set `--enable-advanced-branching` and `--enable-atrig-erf`.

Compilation now fails [1]:
```
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_function.h:376:22: error: cannot convert ‘adub’ to ‘bool’ in return

       { return __x > __y; }
                      ^~~
In file included from /.../dealii/include/deal.II/differentiation/ad/sacado_product_types.h:21:0,
                 from /.../dealii/source/base/symmetric_tensor.cc:20:
/.../dealii/include/deal.II/base/symmetric_tensor.h: In instantiation of ‘bool dealii::internal::SymmetricTensorImplementation::{anonymous}::SortEigenValuesVectors<dim, Number>::operator()(const EigValsVecs&, const EigValsVecs&) [with int dim = 2; Number = adouble; dealii::internal::SymmetricTensorImplementation::{anonymous}::SortEigenValuesVectors<dim, Number>::EigValsVecs = std::pair<adouble, dealii::Tensor<1, 2, adouble> >]’:
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/predefined_ops.h:143:18:   required from ‘constexpr bool __gnu_cxx::__ops::_Iter_comp_iter<_Compare>::operator()(_Iterator1, _Iterator2) [with _Iterator1 = std::pair<adouble, dealii::Tensor<1, 2, adouble> >*; _Iterator2 = std::pair<adouble, dealii::Tensor<1, 2, adouble> >*; _Compare = dealii::internal::SymmetricTensorImplementation::{anonymous}::SortEigenValuesVectors<2, adouble>]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_algo.h:81:17:   required from ‘void std::__move_median_to_first(_Iterator, _Iterator, _Iterator, _Iterator, _Compare) [with _Iterator = std::pair<adouble, dealii::Tensor<1, 2, adouble> >*; _Compare = __gnu_cxx::__ops::_Iter_comp_iter<dealii::internal::SymmetricTensorImplementation::{anonymous}::SortEigenValuesVectors<2, adouble> >]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_algo.h:1921:34:   required from ‘_RandomAccessIterator std::__unguarded_partition_pivot(_RandomAccessIterator, _RandomAccessIterator, _Compare) [with _RandomAccessIterator = std::pair<adouble, dealii::Tensor<1, 2, adouble> >*; _Compare = __gnu_cxx::__ops::_Iter_comp_iter<dealii::internal::SymmetricTensorImplementation::{anonymous}::SortEigenValuesVectors<2, adouble> >]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_algo.h:1953:38:   required from ‘void std::__introsort_loop(_RandomAccessIterator, _RandomAccessIterator, _Size, _Compare) [with _RandomAccessIterator = std::pair<adouble, dealii::Tensor<1, 2, adouble> >*; _Size = long int; _Compare = __gnu_cxx::__ops::_Iter_comp_iter<dealii::internal::SymmetricTensorImplementation::{anonymous}::SortEigenValuesVectors<2, adouble> >]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_algo.h:1968:25:   required from ‘void std::__sort(_RandomAccessIterator, _RandomAccessIterator, _Compare) [with _RandomAccessIterator = std::pair<adouble, dealii::Tensor<1, 2, adouble> >*; _Compare = __gnu_cxx::__ops::_Iter_comp_iter<dealii::internal::SymmetricTensorImplementation::{anonymous}::SortEigenValuesVectors<2, adouble> >]’
/usr/lib/gcc/x86_64-pc-linux-gnu/7.3.0/include/g++-v7/bits/stl_algo.h:4868:18:   required from ‘void std::sort(_RAIter, _RAIter, _Compare) [with _RAIter = std::pair<adouble, dealii::Tensor<1, 2, adouble> >*; _Compare = dealii::internal::SymmetricTensorImplementation::{anonymous}::SortEigenValuesVectors<2, adouble>]’
/.../dealii/include/deal.II/base/symmetric_tensor.templates.h:955:12:   required from ‘std::array<std::pair<Number, dealii::Tensor<1, dim, Number> >, std::integral_constant<int, _Num>::value> dealii::eigenvectors(const dealii::SymmetricTensor<2, dim, Number>&, dealii::SymmetricTensorEigenvectorMethod) [with int dim = 2; Number = adouble]’
source/base/symmetric_tensor.inst:159:40:   required from here
```

```
/.../dealii/include/deal.II/base/tensor.h:836:27: error: cannot convert ‘adub’ to ‘bool’ in return

/.../dealii/include/deal.II/base/tensor.h: In instantiation of ‘bool dealii::Tensor<0, dim, Number>::operator==(const dealii::Tensor<0, dim, OtherNumber>&) const [with OtherNumber = adouble; int dim = 6; Number = adouble]’:
/.../dealii/include/deal.II/base/tensor.h:845:20:   required from ‘bool dealii::Tensor<0, dim, Number>::operator!=(const dealii::Tensor<0, dim, OtherNumber>&) const [with OtherNumber = adouble; int dim = 6; Number = adouble]’
/.../dealii/include/deal.II/base/tensor.h:1161:19:   required from ‘bool dealii::Tensor<rank_, dim, Number>::operator==(const dealii::Tensor<rank_, dim, OtherNumber>&) const [with OtherNumber = adouble; int rank_ = 1; int dim = 6; Number = adouble]’
/.../dealii/include/deal.II/base/symmetric_tensor.h:1418:15:   required from ‘bool dealii::SymmetricTensor<rank, dim, Number>::operator==(const dealii::SymmetricTensor<rank, dim, Number>&) const [with int rank_ = 2; int dim = 3; Number = adouble]’
```

[1] https://cdash.kyomu.43-1.org/buildSummary.php?buildid=1085

### PR DESCRIPTION
**Pull request by jppelteret**
Fixes #6339

### CODE REVIEW

**Comment by tamiko** [comment]
Quickly checking on `tester` (aka bob).

**Comment by jppelteret** [comment]
@tamiko This is a WIP. I'll need some more time to finish adding more patches.

**Comment by jppelteret** [comment]
@tamiko There's still more to go, so I'll have to carry on working on this tomorrow / on the weekend. I'll keep you posted.

**Comment by tamiko** [comment]
@jppelteret Roger, no worries. I will reset the testsuite to non-advanced branching.

File: include/deal.II/base/numbers.h [hunk 1]
```diff
@@ -192,6 +192,29 @@ namespace numbers
...
+   * types (e.g. some automatically differentiable numbers), this function
+   * returns only whether the scalar values stored by the input values are
+   * equal.
+   *
+   * @note This function expects that @p value_2 is castable to the type
+   * of @p value_1.
+   */
+  template <typename Number1,typename Number2>
+  bool
+  are_equal_valued (const Number1 &value_1, const Number2 &value_2);
```
Comment (bangerth) [inline include/deal.II/base/numbers.h]: You can enforce the requirement stated in the documentation above in the function signature if you write it as
```
template <typename Number>
bool
￼ +  are_equal_valued (const Number &value_1, const typename identity<Number>::type &value_2);
```
This way, the compiler will infer `Number` from the first argument, and the second argument will be cast to the same type in the call already. The call will not compile if the second argument cannot be cast to the first argument's type.

File: include/deal.II/base/numbers.h [hunk 2]
```diff
@@ -542,6 +565,51 @@ namespace internal
...
+  bool
+  are_equal_valued (const adouble &value_1,
+                    const adouble &value_2);
+
+  template <typename Number>
+  bool
+  are_equal_valued (const adouble &value_1,
+                    const Number  &value_2)
+  {
+    // Use the specialised definition for two Adol-C taped types
```
Comment (bangerth) [inline include/deal.II/base/numbers.h]: -> specialized

File: include/deal.II/base/numbers.h [hunk 3]
```diff
@@ -542,6 +565,51 @@ namespace internal
...
 }
 
+namespace numbers
+{
+
+#ifdef DEAL_II_ADOLC_WITH_ADVANCED_BRANCHING
+  // Defined in differentiation/ad/adolc_number_types.cc
+  bool
+  are_equal_valued (const adouble &value_1,
+                    const adouble &value_2);
```
Comment (bangerth) [inline include/deal.II/base/numbers.h]: Do these functions show up in the doxygen index? If so, you'll want to document them.

File: include/deal.II/base/numbers.h [hunk 4]
```diff
@@ -542,6 +565,51 @@ namespace internal
...
+    return are_equal_valued(value_2, value_1);
+  }
+#endif
+
+
+  template <typename Number1,typename Number2>
+  bool
+  are_equal_valued (const Number1 &value_1, const Number2 &value_2)
+  {
+    return (value_1 == internal::NumberType<Number1>::value(value_2));
```
Comment (bangerth) [inline include/deal.II/base/numbers.h]: I see that you don't just do a `static_cast`. This may make the option I outlined above infeasible, but at least, it would be worth stating in the documentation how exactly the cast of the second argument is supposed to happen.

File: include/deal.II/base/symmetric_tensor.templates.h [hunk 5]
```diff
@@ -42,7 +42,9 @@ std::array<Number,2>
 eigenvalues (const SymmetricTensor<2,2,Number> &T)
 {
   const Number upp_tri_sq = T[0][1]*T[0][1];
-  if (upp_tri_sq == Number(0.0))
+  // TODO [JPP]: This is not the correct way to do branching with Adol-C
+  // taped doubles with advanced branching is enabled.
+  if (numbers::is_zero_valued(upp_tri_sq))
```
Comment (bangerth) [inline include/deal.II/base/symmetric_tensor.templates.h]: Is this (and the places below) the right place to leave this TODO? It seems pretty unrelated to the purpose of the function. If you think that there is something wrong with this sequence here, maybe the right place for the TODO is inside `is_zero_valued()`...

File: source/fe/fe_values.cc [hunk 6]
```diff
@@ -97,6 +97,36 @@ namespace internal
...
+
+  namespace
+  {
+    // Check to see if a DoF value is zero, implying that subsequent operations
+    // with the value have no effect.
+    template <typename Number, typename T = void>
+    struct FunctionDoFValues
+    {
+      static bool
+      filter (const Number &value)
```
Comment (bangerth) [inline source/fe/fe_values.cc]: `filter` is an awkward name for the function. Filtering means to selectively choose elements out of a sequence, but here you have a *predicate* that only returns true/false. The struct name `FunctionDoFValues` is also not very helpful. Maybe use `CheckForZero` and `CheckForZero::check()`?

File: source/fe/fe_values.cc [hunk 7]
```diff
@@ -489,9 +519,8 @@ namespace FEValuesViews
             // For auto-differentiable numbers, the fact that a DoF value is zero
             // does not imply that its derivatives are zero as well. So we
             // can't filter by value for these number types.
-            if (!Differentiation::AD::is_ad_number<Number>::value)
-              if (value == dealii::internal::NumberType<Number>::value(0.0) )
-                continue;
+            if (dealii::internal::FunctionDoFValues<Number>::filter(value) == true)
+              continue;
```
Comment (bangerth) [inline source/fe/fe_values.cc]: good catch! this is indeed a bug!

**Comment by jppelteret** [comment]
@davydden I don't want to run tests on this while I'm still developing the PR.

**Comment by davydden** [comment]
i guess this also blocks #6156

**Comment by tamiko** [comment]
@jppelteret Bob is happy with the current version https://cdash.kyomu.43-1.org/buildSummary.php?buildid=1157

**Comment by tamiko** [comment]
@jppelteret Bob is happy with the current version https://cdash.kyomu.43-1.org/buildSummary.php?buildid=1159

**Comment by jppelteret** [comment]
@tamiko @bangerth I think I've added as much as I need here. If you're willing to, can you please take another look at this?

**Comment by tamiko** [comment]
@jppelteret Will have a look this evening!

**Comment by tamiko** [comment]
Compiling again on my machine.
